File: android/app/src/main/java/com/example/fmeexample/vm/SmartBotViewModel.kt
/*
 * Copyright (c) 2024-2025 Wingify Software Pvt. Ltd.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */
package com.example.fmeexample.vm

import android.content.Context
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.fmeexample.BuildConfig
import com.example.fmeexample.analytics.MixpanelIntegration
import com.example.fmeexample.fme.FmeHelper
import com.example.fmeexample.models.ChatResponse
import com.example.fmeexample.models.UserInfo
import com.example.fmeexample.useCases.SendChatQueryUseCaseImpl
import com.vwo.interfaces.integration.IntegrationCallback
import kotlinx.coroutines.launch

class SmartBotViewModel(private val chatQueryUseCase: SendChatQueryUseCaseImpl) : ViewModel() {
    private val _chatResponse = MutableLiveData<ChatResponse>()
    val chatResponse: LiveData<ChatResponse> = _chatResponse

    private val _userInfo = MutableLiveData<UserInfo>()
    val userInfo: LiveData<UserInfo> = _userInfo

    private var mixpanelIntegration: MixpanelIntegration? = null

    suspend fun init(context: Context) {
        // Get Mixpanel token from BuildConfig
        val mixpanelToken = BuildConfig.MIXPANEL_PROJECT_TOKEN

        val integrations = if (mixpanelToken.isNotEmpty()) {
            // Initialize Mixpanel
            mixpanelIntegration = MixpanelIntegration.getInstance(context, mixpanelToken)

            object : IntegrationCallback {
                override fun execute(properties: Map<String, Any>) {
                    // Check if this is a flag evaluation or event tracking
                    if (properties["api"] == "track") {
                        // This is event tracking
                        val eventName = properties["eventName"] as String
                        mixpanelIntegration?.trackEvent(eventName, properties)

                    } else if (properties.containsKey("featureName")) {
                        // This is a flag evaluation
                        mixpanelIntegration?.trackFlagEvaluation(properties)
                    }
                }
            }
        } else {
            null
        }
        loadInitialData()
        initFme(context, integrations)
    }

    private fun loadInitialData() {
        viewModelScope.launch {
            _userInfo.value = UserInfo("", "")
        }
    }

    fun sendQuery(userId: String, query: String) {
        viewModelScope.launch {
            processUserQuery(userId, query)
        }
    }

    fun getUserId() {
        viewModelScope.launch {
            _userInfo.postValue(UserInfo(generateRandomUserId(), generateRandomQuery()))
        }
    }

    private fun generateRandomUserId(): String {
        val letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
        return (1..12)
            .map { letters.random() }
            .joinToString("")
    }

    private fun generateRandomQuery(): String {
        return "I forgot my password"
    }

    private suspend fun initFme(context: Context, integrations: IntegrationCallback?) {
        FmeHelper.initFME(context, integrations)
    }

    private suspend fun processUserQuery(userId: String, query: String) {
        val chatResponse = chatQueryUseCase.getChatResponse(userId, query)
        if (chatResponse != null) {
            _chatResponse.postValue(chatResponse)
        }
    }

    fun sendEvent(userId: String) {
        chatQueryUseCase.sendEvent(userId)
    }
}

File: flutter/lib/vm/SmartBotViewModel.dart
/*
 * Copyright (c) 2025 Wingify Software Pvt. Ltd.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:vwo_fme_flutter_sdk_example/fme/fme_service.dart';

import '../models/ChatResponse.dart';
import '../models/UserInfo.dart';
import '../usecases/ISendChatQueryUseCase.dart';

class SmartBotViewModel with ChangeNotifier {
  final ISendChatQueryUseCase chatQueryUseCase;

  UserInfo? _userInfo;
  UserInfo? get userInfo => _userInfo;

  ChatResponse? _chatResponse;
  ChatResponse? get chatResponse => _chatResponse;

  SmartBotViewModel(this.chatQueryUseCase);

  Future<void> init(BuildContext context) async {
    await loadInitialData();
    await initFme();
  }

  Future<void> loadInitialData() async {
    _userInfo = UserInfo(userId: "", query: "");
    notifyListeners();
  }

  Future<void> sendQuery(String userId, String query) async {
    await processUserQuery(userId, query);
  }

  Future<void> getUserId() async {
    _userInfo = UserInfo(userId: _generateRandomUserId(), query: _generateRandomQuery());
    notifyListeners();
  }

  setUserId(String userId) async {
    _userInfo = UserInfo(userId: userId, query: _generateRandomQuery());
    notifyListeners();
  }

  String _generateRandomUserId() {
    const letters = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
    return List.generate(12, (index) => letters[Random().nextInt(letters.length)]).join();
  }

  String _generateRandomQuery() {
    return 'I forgot my password';
  }

  Future<void> initFme() async {
    FmeService().init();
  }

  Future<void> processUserQuery(String userId, String query) async {
    final response = await chatQueryUseCase.getChatResponse(userId, query);
    if (response != null) {
      _chatResponse = response;
      notifyListeners();
    }
  }

  Future<void> sendEvent(String userId) async {
    await chatQueryUseCase.sendEvent(userId);
  }

  String getFlagStatus() {
    return chatResponse?.isEnabled == true ? "Enabled" : "Disabled";
  }
}